infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/query/impl/ProtostreamPayloadClosableIterator.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.ogm.datastore.infinispanremote.query.impl;
+
+import java.util.List;
+
+import org.hibernate.ogm.datastore.infinispanremote.impl.protostream.ProtostreamPayload;
+import org.hibernate.ogm.dialect.query.spi.ClosableIterator;
+import org.hibernate.ogm.model.spi.Tuple;
+
+/**
+ * Iterates over the result of an Infinispan query, when each result is a **full** cache entry.
+ * This is the case when the result of query is mapped from a Protostream payload type.
+ *
+ * @author Fabio Massimo Ercoli
+ */
+public class ProtostreamPayloadClosableIterator implements ClosableIterator<Tuple> {
+
+	private final List<ProtostreamPayload> queryResult;
+	private int index = 0;
+
+	public ProtostreamPayloadClosableIterator(List<ProtostreamPayload> queryResult) {
+		this.queryResult = queryResult;
+	}
+
+	@Override
+	public boolean hasNext() {
+		return index < queryResult.size();
+	}
+
+	@Override
+	public Tuple next() {
+		return queryResult.get( index++ ).toTuple( Tuple.SnapshotType.UPDATE );
```

Comment:
It does not return null if the method was invoked only after hasNext. Checking the index could be defensive programming, I think. In my opinion defensive programming should be applied only to API, or SPI, not for internal collaboration. But as you prefer.

Reply:
What I'm asking is if the result could be 10 null elements? I don't think it would happen but I'm checking with you.
Again, it is not about what I prefer, I'm asking you if that use case can happen

Reply:
Sorry I see only now the point. I'm going to check if it possible with a test.

Reply:
Yes it is possible to store null value. But is not possible having its in query result set, even if it was a bulk get. See [Test|https://github.com/fax4ever/infinispan-play/blob/master/infinispan-query-play/src/test/java/it/redhat/demo/test/QueryIT.java]::getBulk_simple.

Reply:
OK, thanks. Just wanted to double check